src/core/qgsvectorlayer.cpp

```diff
@@ -3923,6 +3918,14 @@ QVariant QgsVectorLayer::aggregate( QgsAggregateCalculator::Aggregate aggregate,
     return QVariant();
   }
 
+  if ( context && !fids )
+  {
+    if ( context->indexOfScope( "Symbol scope" ) != -1 )
+    {
+      fids = &mSymbolIdMap.value( context->variable( "symbol_id" ).toString(), QgsFeatureIds() );
```

Comment:
You will have to find another solution for this. Making the aggregate function depend on specific context variables is unexpected behaviour.

As discussed previously: I would propose to save the map in the feature counter and obtain the fids directly from there and then pass the fids to this aggregate function.

Reply:
@m-kuhn I have nothing against using the feature counter, but we went over the fact that this causes segfault constantly.

We already went over this point and I never obtained a reply after demonstrating that the counter is unreliable even in the current API.

For the scope part we already discussed this also, I cannot find another way to inject a variable between the expression parsing and the call to aggregate. If you know of a way please say so. Or if you have another alternative please say so. Otherwise this PR was a waste of time.

Reply:
I have removed the scope check, is that ok?

As for the featurecounter, if you want to rely on it, you would have to recount everything for each evaluation as it cannot be used as a storage.

Reply:
I remember past discussions about this topic. I can't find the discussion again right now, but I remember that the following suggestions were made: connect to the `symbolsCounted` signal and directly extract the information from there. That should not result in a segfault since it's using the symbol counter the way it's designed to be used. Caching was also mentioned in this discussion. I think caching could be worth a discussion, but will need to be discussed on a broader level (cache invalidation, caching of other feature count results) which I would prefer not to have inside this pull request in order to get it finally ready to be merged :)

Reply:
the part about the symbolCounted, that was for to trigger the refreshing of the legend to prevent freezing qgis ( this is why the current test is failing as the snapshot is saved before the features are counted.

Currently there are two issues at hand that are still lingering:

1. How to input the symbol id to obtain the associated filtered fids.
2. How should those fids be cached.

And here is my current standing on those issues:

1. If I cannot rely on the present of the id variable in the scope, the only other alternative between the evaluateExpression and the qgsvectoraggregate is either to check the scope for the variable (I understand that if the symbol scope would be reused with an aggregate maybe, just maybe people wouldn't want to use only the items in the symbol scope), but I am assuming 2 things on the other hand: 1: the symbol is only currently used for this end and it should not be an issue, 2 if people were to use the symbol scope for any sort of calculation, I would assume that they are interested in a substed of the layer that is represented by a particular symbol and that the current behaviour would be greatly benefic.
Otherwise to be able to pass the symbol in the aggregate functions, I would have to inject something in the aggregate function and all other related function in order to be able to filter the data by a symbol id, this entails that I need to change every aggregate function to be able to accept a new argument and ensure that it is handled properly.
2.  I have tested extensively the use of the featurecounter and it only provided segfaults. This is also the reason why the feature count (which is smaller that the fids registry) is transferred to the qgsvectorlayer rather than the counter itself. 
Here is the suggested test, to prove that even basic usage of the counter inside the vectorlayer would be problematic
```python
layer = iface.activeLayer()
counter=layer.countSymbolFeatures()
counter.featureCount('1')
```
For more proof, please see https://travis-ci.org/qgis/QGIS/jobs/521792574 and every other tests results done around the 18th of april. I have tested everything that I could to ensure that the qgsfeatureids was existing but it always get deleted quickly after being created, this lead to segfault and explains all the safety that I have tried during that period.

Reply:
Thanks for the thorough comments:

1. First thing: QgsVectorLayer is deep in the core library, if it was in app there would be no issue at all. We are all trying to make QGIS core a pleasant to use API that behaves predictable. So even if it works for this feature here, it will act as a precedent for more similar approaches introduced and if we ever get a "proper" API for this, we will break code people wrote who depended on this implementation. So ideally, this will be based on the outcome of the discussion here https://github.com/qgis/QGIS/pull/9915#issuecomment-505734932

2. Caching this in the vector layer has some tricky issues too like for example using a different symbology for the print map than the current canvas "main" symbology. Not that the example below would support that yet, but with the proposed code below we keep the door open at least:

```
layer = iface.activeLayer()
counter = QgsVectorLayerFeatureCounter(layer)

def onFeaturesCounted():
    print('Counted {} features'.format(counter.featureCount('1')))

counter.symbolsCounted.connect(onFeaturesCounted)
QgsApplication.taskManager().addTask( counter );
```

Reply:
Thanks for the reply

1. I totally agree that this depending on the filtering approach. I know that this can set a precedent but you must also understood that I have exhausted all alternative solutions that I could currently come up with. I will wait and see how the filtering proposal evolve. In the even that the proposed exchange takes too long to occur or simply never occur, could it be possible to have this merged anyway and this approach implemented as a band-aid until those discussion are over? I doubt this will affect other people's code as the symbol expression context is a total novelty that is not used, so people would not fall for the proposed approach.

2. I totally agree that this might be an issue, but then again this is the case with the counter. The current proposal does not derail from current API code. I am also unsure how common it is to have distinct symbology in the canvas and in the map as I understand that the canvas symbology is imported from the layertree. I'm not sure how much we should prepare to have people combine two fringe cases (distinct canvas symbology, if that is even possible + legend symbol expression). Otherwise hosting and maintaining a counter in the canvas or in the layer itself seem like the same amount of memory to me. Otherwise the comment that I made yesterday is still valid.

Reply:
1. You can always use an expression based filter with ids as I suggested. That won't include any API change in QgsVectorLayer. In the future we can discuss the details of the API change to improve the performance.
2. Not sure what else I can say.